### PR TITLE
Batched Matrix Multiplication Operation

### PR DESCRIPTION
**Pull request by ozanarkancan**
Hi,

This is a premature pr for the #295. I simply made an interface for the [gemmStridedBatched](https://devblogs.nvidia.com/cublas-strided-batched-matrix-multiply/) call. All comments are welcome.

```julia
#size of x1: (m,k,bs)
#size of x2: (k,n,bs)
#size of output: (m,n,bs)
output = bmm(x1, x2)
```

### CODE REVIEW

File: src/bmm.jl [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    end
+    return C
+end
+
+function bmm!(A, B, C)
+    m,n,bs = size(A)
+    n,k,bs = size(B)
+    
+    for i=1:bs
+        C[:,:,i] = reshape(A[:,:,i], m, n) * reshape(B[:, :, i], n, k)
```
Comment (ekinakyurek) [inline src/bmm.jl]: may be it is good to use `view(A,:,:,i)` here. I think we should have tests for `bmm` with profiling to compare different methods.

**Comment by ngphuoc** [comment]
Thanks a lot for the PR. I got 100 times speedup in the following benchmark!

```julia
using Knet, BenchmarkTools

m,n,p,b = 100,200,300,50
atype = KnetArray{Float32}
a = rand(m,n,b)
b = rand(n,p,b)
a_ = atype(a)
b_ = atype(b)

@btime for i=1:size(a,3)
  a[:,:,i] * b[:,:,i]
end;  # 11.323 ms (651 allocations: 41.98 MiB)

@btime for i=1:size(a,3)
  a_[:,:,i] * b_[:,:,i]
end;  # 338.570 μs (1999 allocations: 64.06 KiB)

@btime c = bmm(a, b);  # 12.012 ms (302 allocations: 53.42 MiB)

@btime c_ = bmm(a_, b_);  # 3.822 μs (7 allocations: 336 bytes)
```julia

File: src/bmm.jl [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+function bmm!(A, B, C)
+    if ndims(A) != 3 || ndims(B) != 3
+        throw(DimensionMismatch("$(map(size,(A,B,C)))"))
+    end
+    ma,ka,bsa = size(A)
+    kb,nb,bsb = size(B)
+
+    (bsa == bsb && ka == kb && size(C,1) == ma && size(C,2) == nb && size(C,3) == bsa) || throw(DimensionMismatch("$(map(size,(A,B,C)))"))
+
```
Comment (ngphuoc) [inline src/bmm.jl]: I need to add this
```suggestion
    m,n,k,bs = ma,nb,ka,bsa
```

**Comment by denizyuret** [comment]
I assume this is ready to go?

**Comment by ozanarkancan** [comment]
Yes, this can be merged.

**Comment by ilkerkesen** [comment]
Resolved the merge conflict and fixed cuda macro call. Tests pass but I don't now anything about the table in `runtests.jl`. It's working and ready to be merged. @denizyuret

**Pull request merged**